src/cargo/core/compiler/fingerprint/dirty_reason.rs

```diff
@@ -0,0 +1,508 @@
+
+                Ok(())
+            }
+            DirtyReason::NothingObvious => {
+                // See comment in fingerprint compare method.
+                write!(f, "two fingerprint comparison turned up nothing obvious")
+            }
+            DirtyReason::Forced => write!(f, "forced"),
+        }
+    }
+}
+
+impl std::error::Error for DirtyReason {}
+
+pub struct DirtyReasonPrettyPresentation<'a>(&'a DirtyReason, &'a Unit, &'a Config);
+
+trait ShellExt {
+    fn dirty_because(&mut self, unit: &Unit, s: impl fmt::Display) -> CargoResult<()>;
+}
+
+impl ShellExt for Shell {
+    fn dirty_because(&mut self, unit: &Unit, s: impl fmt::Display) -> CargoResult<()> {
+        self.status("Dirty", format!("{}: {s}", &unit.pkg))
+    }
+}
+
+struct FileTimeDiff {
+    old_time: FileTime,
+    new_time: FileTime,
+}
+
+impl fmt::Display for FileTimeDiff {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        let s_diff = self.new_time.seconds() - self.old_time.seconds();
+        if s_diff >= 1 {
+            let mins_diff = s_diff / 60;
+
+            if mins_diff >= 1 {
+                let s_diff = s_diff % 60;
+
+                let hours_diff = mins_diff / 60;
+                if hours_diff >= 1 {
+                    let mins_diff = mins_diff % 60;
+
+                    write!(f, "{hours_diff}h {mins_diff}min {s_diff}s")
+                } else {
+                    write!(f, "{mins_diff}min {s_diff}s")
+                }
+            } else {
+                write!(f, "{s_diff}s")
+            }
+        } else {
+            let ns_diff = self.new_time.nanoseconds() - self.old_time.nanoseconds();
+            write!(f, "{ns_diff}ns")
+        }
+    }
+}
+
+struct After {
+    old_time: FileTime,
+    new_time: FileTime,
+    what: &'static str,
+}
+
+impl fmt::Display for After {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}, {} after {} at {}",
+            self.new_time,
+            FileTimeDiff {
+                old_time: self.old_time,
+                new_time: self.new_time
+            },
+            self.what,
+            self.old_time
+        )
+    }
+}
+
+impl<'a> DirtyReasonPrettyPresentation<'a> {
+    fn present_dependency_diffs(
+        s: &mut Shell,
+        unit: &Unit,
+        _deps: &DirtyReasonDeps,
+    ) -> CargoResult<()> {
+        s.dirty_because(unit, "the list of dependencies changed")?;
+
+        Ok(())
+    }
+
+    fn after(old_time: FileTime, new_time: FileTime, what: &'static str) -> After {
+        After {
+            old_time,
+            new_time,
+            what,
+        }
+    }
+
+    pub fn present_to(&self, s: &mut Shell) -> CargoResult<()> {
+        let unit = self.1;
+        let ws_roots = self.2.ws_roots.borrow();
+        let root: &Path = ws_roots
+            .values()
+            .next()
+            .map_or_else(|| self.2.cwd(), |it| it.inheritable().ws_root().as_ref());
```

Comment:
Is this the right way to find the root of the workspace? (It seemed to work while I was testing it.)

Reply:
I would get it from maybr `cx.bcx.ws.root()`. But maybe we should just print the full path and the user doesn't need to figure out whether it is workspace root, package root, or current directory.